libraries/chain/thread_utils.cpp

```diff
@@ -23,10 +24,33 @@ void named_thread_pool::start() {
    _ioc_work.emplace( boost::asio::make_work_guard( _ioc ) );
    _ioc.restart();
    for( size_t i = 0; i < _num_threads; ++i ) {
-      _thread_pool.emplace_back( [&ioc = _ioc, &name_prefix = _name_prefix, i]() {
+      _thread_pool.emplace_back( [&ioc = _ioc, &name_prefix = _name_prefix, on_except = _on_except, i]() {
          std::string tn = name_prefix + "-" + std::to_string( i );
-         fc::set_os_thread_name( tn );
-         ioc.run();
+         try {
+            fc::set_os_thread_name( tn );
+            ioc.run();
+         } catch( const fc::exception& e ) {
+            if( on_except ) {
+               on_except( e );
+            } else {
+               elog( "Exiting thread ${t} on exception: ${e}", ("t", tn)("e", e.to_detail_string()) );
+            }
+         } catch( const std::exception& e ) {
```

Comment:
Is it on purpose that you change the exiting behavior by catching the exceptions and not rethrowing them (even when `on_except` is empty).
When `on_except` is not empty, I guess the way you use it by calling `app().quit()` makes it OK to not rethrow, but should we assume that will always be the case? and if that is always the case, maybe instead of passing a lambda, it would be simpler to pass an "origin" string and call `app().quit()` here?
Just some thoughts!

Reply:
Hmm, yeah, maybe re-throw is better when not provided an `on_except`. If provided an `on_except` the user can throw or not depending on what they want to do.

`app()` is not available everywhere `thread_pool` is used.